File: apps/web/src/lib/api/inventory-assignment.ts
import { apiFetch } from '@/lib/api/client';

export const ASSIGNMENT_TYPES = ['parking_for', 'storage_for'] as const;
export type AssignmentType = (typeof ASSIGNMENT_TYPES)[number];

export const ASSIGNMENT_REQUEST_TYPES = ['assign', 'reassign', 'unassign'] as const;
export type AssignmentRequestType = (typeof ASSIGNMENT_REQUEST_TYPES)[number];

export const ASSIGNMENT_REQUEST_STATUSES = [
  'draft',
  'submitted',
  'under_review',
  'approved',
  'rejected',
  'revision_requested',
  'withdrawn',
  'applied',
  'stale',
] as const;

export type AssignmentRequestStatus = (typeof ASSIGNMENT_REQUEST_STATUSES)[number];

export interface AssignmentRecord {
  id: string;
  child_asset_id: string;
  child_display_id: string | null;
  parent_asset_id: string;
  parent_display_id: string | null;
  assignment_type: AssignmentType;
  status: string;
  effective_from: string;
  effective_to: string | null;
  assignment_price: string | null;
  currency: string;
  supporting_document_id: string | null;
  related_transaction_id: string | null;
  notes: string | null;
  created_by_user_id: string | null;
  approved_by_user_id: string | null;
  assignment_request_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface AssignmentRequest {
  id: string;
  child_asset_id: string;
  child_display_id: string | null;
  parent_asset_id: string | null;
  parent_display_id: string | null;
  project_name: string | null;
  request_type: AssignmentRequestType;
  assignment_type: AssignmentType;
  effective_date: string;
  reason: string;
  assignment_price: string | null;
  currency: string;
  supporting_document_id: string | null;
  related_transaction_id: string | null;
  requested_by_user_id: string;
  assigned_approver_user_id: string | null;
  status: AssignmentRequestStatus;
  reviewed_at: string | null;
  approved_at: string | null;
  rejected_at: string | null;
  decision_notes: string | null;
  actions: string[];
  created_at: string;
  updated_at: string;
}

export interface AssignmentSummary {
  assigned_to_display_id: string | null;
  assigned_to_asset_id: string | null;
  assignment_status: string;
  assignment_type: string | null;
  parking_count: number;
  storage_count: number;
  has_scheduled: boolean;
  has_pending: boolean;
}

export interface AssetAssignmentDetail {
  child_asset_id: string;
  current: AssignmentRecord | null;
  history: AssignmentRecord[];
  pending_requests: AssignmentRequest[];
  scheduled_requests: AssignmentRequest[];
  parent_accessories: AssignmentRecord[];
}

export interface AssignmentRequestInput {
  child_asset_id: string;
  parent_asset_id?: string | null;
  request_type?: AssignmentRequestType;
  effective_date: string;
  reason: string;
  assignment_price?: string | null;
  currency?: string;
  supporting_document_id?: string | null;
  related_transaction_id?: string | null;
  assigned_approver_user_id?: string | null;
  submit?: boolean;
}

export function fetchAssetAssignmentDetail(assetId: string): Promise<AssetAssignmentDetail> {
  return apiFetch(`/inventory/assignments/by-asset/${assetId}/detail`);
}

export function fetchAssignmentSummary(assetId: string): Promise<AssignmentSummary> {
  return apiFetch(`/inventory/assignments/summary/by-asset/${assetId}`);
}

export function fetchPendingAssignmentApprovals(): Promise<AssignmentRequest[]> {
  return apiFetch('/inventory/assignment-requests/pending-approvals');
}

export function createAssignmentRequest(input: AssignmentRequestInput): Promise<AssignmentRequest> {
  return apiFetch('/inventory/assignment-requests', {
    method: 'POST',
    body: JSON.stringify(input),
  });
}

export function approveAssignmentRequest(requestId: string, comments?: string): Promise<AssignmentRequest> {
  return apiFetch(`/inventory/assignment-requests/${requestId}/approve`, {
    method: 'POST',
    body: JSON.stringify({ comments: comments ?? null }),
  });
}

export function rejectAssignmentRequest(requestId: string, decisionNotes: string): Promise<AssignmentRequest> {
  return apiFetch(`/inventory/assignment-requests/${requestId}/reject`, {
    method: 'POST',
    body: JSON.stringify({ decision_notes: decisionNotes }),
  });
}

export function requestAssignmentRevision(requestId: string, decisionNotes: string): Promise<AssignmentRequest> {
  return apiFetch(`/inventory/assignment-requests/${requestId}/request-revision`, {
    method: 'POST',
    body: JSON.stringify({ decision_notes: decisionNotes }),
  });
}

export function formatAssignmentMoney(
  value: string | null | undefined,
  currency: string,
  locale: string,
): string {
  if (value === null || value === undefined || value === '') return '—';
  const num = Number.parseFloat(value);
  if (Number.isNaN(num)) return '—';
  return new Intl.NumberFormat(locale, { style: 'currency', currency }).format(num);
}

File: apps/api/src/investhome_api/schemas/inventory_assignment.py
"""Inventory assignment API schemas."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from investhome_api.models.inventory import (
    AssignmentApprovalDecision,
    AssignmentRecordStatus,
    AssignmentRequestStatus,
    AssignmentRequestType,
    AssignmentType,
)


class AssignmentRequestCreate(BaseModel):
    child_asset_id: UUID
    parent_asset_id: UUID | None = None
    request_type: AssignmentRequestType | None = None
    effective_date: date
    reason: str = Field(min_length=1)
    assignment_price: Decimal | None = Field(default=None, ge=0)
    currency: str = Field(default="USD", min_length=3, max_length=3)
    supporting_document_id: UUID | None = None
    related_transaction_id: UUID | None = None
    assigned_approver_user_id: UUID | None = None
    submit: bool = False


class AssignmentRequestUpdate(BaseModel):
    parent_asset_id: UUID | None = None
    effective_date: date | None = None
    reason: str | None = Field(default=None, min_length=1)
    assignment_price: Decimal | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    supporting_document_id: UUID | None = None
    related_transaction_id: UUID | None = None
    assigned_approver_user_id: UUID | None = None


class AssignmentDecisionInput(BaseModel):
    comments: str | None = None
    decision_notes: str | None = None


class AssignmentRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    child_asset_id: UUID
    child_display_id: str | None = None
    parent_asset_id: UUID
    parent_display_id: str | None = None
    assignment_type: AssignmentType
    status: AssignmentRecordStatus
    effective_from: date
    effective_to: date | None
    assignment_price: Decimal | None
    currency: str
    supporting_document_id: UUID | None
    related_transaction_id: UUID | None
    notes: str | None
    created_by_user_id: UUID | None
    approved_by_user_id: UUID | None
    assignment_request_id: UUID | None
    created_at: datetime
    updated_at: datetime


class AssignmentApprovalRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    reviewer_user_id: UUID
    decision: AssignmentApprovalDecision
    comments: str | None
    created_at: datetime


class AssignmentRequestResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    child_asset_id: UUID
    child_display_id: str | None = None
    parent_asset_id: UUID | None
    parent_display_id: str | None = None
    project_name: str | None = None
    request_type: AssignmentRequestType
    assignment_type: AssignmentType
    effective_date: date
    reason: str
    assignment_price: Decimal | None
    currency: str
    supporting_document_id: UUID | None
    related_transaction_id: UUID | None
    requested_by_user_id: UUID
    assigned_approver_user_id: UUID | None
    status: AssignmentRequestStatus
    reviewed_at: datetime | None
    approved_at: datetime | None
    rejected_at: datetime | None
    decision_notes: str | None
    actions: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class AssignmentRequestListResponse(BaseModel):
    items: list[AssignmentRequestResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class AssignmentSummaryResponse(BaseModel):
    assigned_to_display_id: str | None
    assigned_to_asset_id: str | None
    assignment_status: str
    assignment_type: str | None
    parking_count: int
    storage_count: int
    has_scheduled: bool
    has_pending: bool


class AssetAssignmentDetailResponse(BaseModel):
    child_asset_id: UUID
    current: AssignmentRecordResponse | None
    history: list[AssignmentRecordResponse]
    pending_requests: list[AssignmentRequestResponse]
    scheduled_requests: list[AssignmentRequestResponse]
    parent_accessories: list[AssignmentRecordResponse] = Field(default_factory=list)
